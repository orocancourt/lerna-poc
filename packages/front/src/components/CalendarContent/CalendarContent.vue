<template>
  <v-layout
    id="calendar-content"
    class="table-row"
    row>
    <v-flex
      class="table-column"
      xs3/>
    <v-flex
      v-for="day in numberOfDays"
      :key="day"
      class="table-column"
      xs2
      pa-0>
      <calendar-cell/>
    </v-flex>
  </v-layout>
</template>

<script>
import { mapState } from 'vuex'
import CalendarCell from '../CalendarCell/CalendarCell.vue'

export default {
  name: 'CalendarContent',
  components: { CalendarCell },
  props: {
    numberOfDays: {
      type: Number,
      default: 8,
    },
    numberOfPastDays: {
      type: Number,
      default: 2,
    },
  },
  computed: mapState(['currentDate']),
}
</script>

<style lang="scss" scoped>
</style>
