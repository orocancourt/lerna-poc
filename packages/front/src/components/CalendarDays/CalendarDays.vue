<template>
  <v-layout
    id="calendar-days"
    class="table-row"
    row>
    <v-flex
      class="table-column"
      xs3/>
    <v-btn
      class="calendar-navigation-button"
      icon
      @click="showPreviousDay()">
      <v-icon>keyboard_arrow_left</v-icon>
    </v-btn>
    <v-flex
      v-for="(day, index) in days"
      :key="index"
      class="table-column calendar-day"
      xs2
      pa-0>
      <div
        :class="{
          'current-date': day.isCurrentDate,
          'today': day.isToday,
          'past-day': day.isPastDay,
          'future-day': day.isFutureDay,
        }"
        class="calendar-day-button">
        <span class="calendar-day-button-text">
          <span class="calendar-day-button-text-digit">{{ day.digit }}</span>
          <span class="calendar-day-button-text-weekday">{{ day.weekday }}</span>
        </span>
        <span
          v-if="day.isCurrentDate"
          class="calendar-day-button-underline"/>
      </div>
    </v-flex>
    <v-btn
      class="calendar-navigation-button"
      icon
      @click="showNextDay()">
      <v-icon>keyboard_arrow_right</v-icon>
    </v-btn>
  </v-layout>
</template>

<script>
  import moment, { Moment } from 'moment'
  import { times } from 'lodash'

  export default {
    name: 'CalendarDays',
    props: {
      currentDate: {
        type: Moment,
        default: () => moment(),
      },
      numberOfDays: {
        type: Number,
        default: 8,
      },
      numberOfPastDays: {
        type: Number,
        default: 2,
      },
    },
    computed: {
      today() {
        return moment()
      },
      startDate() {
        return moment(this.currentDate).subtract(this.numberOfPastDays, 'day')
      },
      days() {
        return times(this.numberOfDays, i => {
          const day = moment(this.startDate).add(i, 'day')
          return {
            digit: this.formatDayDigitFromDate(day),
            weekday: this.formatWeekdayFromDate(day),
            isCurrentDate: this.isCurrentDate(day),
            isToday: this.isToday(day),
            isPastDay: this.isPastDay(day),
            isFutureDay: this.isFutureDay(day),
          }
        })
      },
    },
    methods: {
      formatDayDigitFromDate(date) {
        return date.locale('fr').format('D')
      },
      formatWeekdayFromDate(date) {
        return date.locale('fr').format('ddd').slice(0, -1)
      },
      isCurrentDate(date) {
        return date.isSame(this.currentDate, 'day')
      },
      isToday(date) {
        return date.isSame(this.today, 'day')
      },
      isPastDay(date) {
        return date.isBefore(this.today, 'day')
      },
      isFutureDay(date) {
        return date.isAfter(this.today, 'day')
      },
      showPreviousDay() {
        this.currentDate = moment(this.currentDate).subtract(1, 'day')
      },
      showNextDay() {
        this.currentDate = moment(this.currentDate).add(1, 'day')
      },
    },
  }
</script>

<style lang="scss" scoped>
  $rowHeight: 90px;
  $underlineHeight: 4px;
  $navigationButtonSize: 24px;
  $navigationButtonMarginY: ($rowHeight - $navigationButtonSize) / 2;

  .table-row {
    position: sticky;
    top: 70px;
    min-height: $rowHeight;
    max-height: $rowHeight;
  }

  .calendar-navigation-button {
    position: absolute;
    width: $navigationButtonSize;
    height: $navigationButtonSize;
    margin: $navigationButtonMarginY 0;
    color: $lightBlue;

    &:first-of-type {
      left: 15%;
      margin-left: 9px;
      margin-right: 9px;
    }

    &:last-of-type {
      right: 0;
    }
  }

  .calendar-day {
    display: flex;
    flex-direction: column;

    .calendar-day-button {
      display: flex;
      flex-direction: column;
      flex-grow: 1;
      align-items: stretch;
      justify-content: center;

      &.today {
        color: $lightBlue;
        background-color: $lightBlue;
      }
      &.past-day {
        color: $grey;
        background-color: $grey;
      }
      &.future-day {
        color: $darkBlue;
        background-color: $darkBlue;
      }
      &.current-date > span {
        margin-bottom: -$underlineHeight / 2;
      }

      .calendar-day-button-text {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        justify-content: center;
        background-color: white;

        .calendar-day-button-text-digit {
          font-size: 36px;
        }
        .calendar-day-button-text-weekday {
          font-size: 16px;
          margin-top: -10px;
        }
      }

      .calendar-day-button-underline {
        width: 60%;
        height: $underlineHeight;
        margin: 0 auto;
        z-index: 1;
        background-color: inherit;
      }
    }
  }
</style>
