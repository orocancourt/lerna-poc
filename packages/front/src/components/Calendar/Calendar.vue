<template>
  <section id="calendar">
    <div class="calendar-navigation">
      <v-btn
        class="calendar-navigation-button"
        icon
        @click="showPreviousMonth()">
        <v-icon>keyboard_arrow_left</v-icon>
      </v-btn>
      <h1>{{ formattedCurrentDate }}</h1>
      <v-btn
        class="calendar-navigation-button"
        icon
        @click="showNextMonth()">
        <v-icon>keyboard_arrow_right</v-icon>
      </v-btn>
    </div>
    <v-container
      class="table"
      text-xs-center
      pa-0>
      <calendar-days
        :current-date="currentDate"
        :number-of-days="numberOfDays"
        :number-of-past-days="numberOfPastDays"/>
      <calendar-content
        :number-of-days="numberOfDays"
        :number-of-past-days="numberOfPastDays"
      />
    </v-container>
  </section>
</template>

<script>
  import { mapState } from 'vuex'
  import moment from 'moment'
  import CalendarDays from '../CalendarDays/CalendarDays.vue'
  import CalendarContent from '../CalendarContent/CalendarContent.vue'

  export default {
    name: 'Calendar',
    components: { CalendarDays, CalendarContent },
    data: () => ({
      numberOfDays: 8,
      numberOfPastDays: 2,
    }),
    computed: {
      formattedCurrentDate() {
        return this.formatDate(this.currentDate)
      },
      // currentDate() {
      //   console.log('current date called')
      //
      //   if (this.$store) {
      //     return this.$store.state.currentDate
      //   }
      //   return moment()
      // },
      ...mapState(['currentDate']),
    },
    methods: {
      formatDate(date) {
        return date.locale('fr').format('MMMM YYYY')
      },
      showPreviousMonth() {
        this.currentDate = moment(this.currentDate).subtract(1, 'month')
      },
      showNextMonth() {
        this.currentDate = moment(this.currentDate).add(1, 'month')
      },
    },
  }
</script>

<style lang="scss">
  $navigationButtonSize: 24px;

  section {
    height: 100%;
    padding: 20px;
    background-color: $sandGrey;
  }

  .calendar-navigation {
    display: flex;
    align-items: center;
    margin-bottom: 30px;

    .calendar-navigation-button {
      width: $navigationButtonSize;
      height: $navigationButtonSize;
      margin: 0;
      color: $lightBlue;
    }

    h1 {
      width: 224px;
      padding: 10px;
      font-size: 28px;
      font-weight: normal;
      color: $darkBlue;

      &:first-letter {
        text-transform: capitalize;
      }
    }
  }

  .table {
    display: flex;
    flex-direction: column;
    min-height: 70%;
    margin: 10px 0;
    background-color: white;

    .table-row {
      border-bottom: 1px solid $sandGrey;

      &:last-child {
        border-bottom: none;
      }

      .table-column {
        border-left: 1px solid $sandGrey;

        &:first-child {
          border-left: none;
        }
      }
    }
  }
</style>
