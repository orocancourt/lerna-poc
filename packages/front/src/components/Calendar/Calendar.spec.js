import { createLocalVue, mount, shallow } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex from 'vuex'
import moment from 'moment'
import Calendar from './Calendar.vue'
import CalendarDays from '../CalendarDays/CalendarDays.vue'
import CalendarContent from '../CalendarContent/CalendarContent.vue'

describe('components | Calendar', () => {
  let localVue
  let store
  const now = 1487076708000
  const expectedCurrentDate = moment(now)

  beforeEach(() => {
    Date.now = jest.fn(() => now)
    localVue = createLocalVue()
    localVue.use(Vuetify)
    localVue.use(Vuex)
    store = new Vuex.Store({
      state: {
        currentDate: expectedCurrentDate,
      },
    })
  })

  describe('template', () => {
    it('should match snapshot', () => {
      // When
      const wrapper = shallow(Calendar, { store, localVue })

      // Then
      expect(wrapper.element).toMatchSnapshot()
    })

    it('should render calendar-days', () => {
      // When
      const wrapper = mount(Calendar, { store, localVue })

      // Then
      expect(wrapper.find(CalendarDays).element).toBeDefined()
    })

    it('should render calendar-content', () => {
      // When
      const wrapper = mount(Calendar, { store, localVue })

      // Then
      expect(wrapper.find(CalendarContent).element).toBeDefined()
    })

    describe('computed property', () => {
      it('should have currentDate from store', () => {
        // When
        const store = new Vuex.Store({
          state: {
            currentDate: expectedCurrentDate,
          },
        })
        const wrapper = mount(Calendar, { localVue, store })

        // Then
        expect(wrapper.vm.currentDate).toEqual(expectedCurrentDate)
      })
    })

    it('should render current date', () => {
      // Given
      const expectedText = 'novembre 1992'
      const computed = {
        formattedCurrentDate() {
          return expectedText
        },
      }
      // When
      const wrapper = shallow(Calendar, { localVue, computed, store })

      // Then
      expect(wrapper.find('h1').element.textContent).toEqual(expectedText)
    })

    describe('CalendarDays component props', () => {
      it('should set numberOfDays', () => {
        // Given
        const numberOfDays = 5
        const wrapper = mount(Calendar, { store, localVue })

        // When
        wrapper.setData({ numberOfDays })

        // Then
        expect(wrapper.find(CalendarDays).vm.$props.numberOfDays).toEqual(numberOfDays)
      })

      it('should set numberOfPastDays', () => {
        // Given
        const numberOfPastDays = 4
        const wrapper = mount(Calendar, { store, localVue })

        // When
        wrapper.setData({ numberOfPastDays })

        // Then
        expect(wrapper.find(CalendarDays).vm.$props.numberOfPastDays).toEqual(numberOfPastDays)
      })
    })

    describe('CalendarContent component props', () => {
      it('should set numberOfDays', () => {
        // Given
        const numberOfDays = 6
        const wrapper = mount(Calendar, { store, localVue })

        // When
        wrapper.setData({ numberOfDays })

        // Then
        expect(wrapper.find(CalendarContent).vm.$props.numberOfDays).toEqual(numberOfDays)
      })

      it('should set numberOfPastDays', () => {
        // Given
        const numberOfPastDays = 1
        const wrapper = mount(Calendar, { store, localVue })

        // When
        wrapper.setData({ numberOfPastDays })

        // Then
        expect(wrapper.find(CalendarContent).vm.$props.numberOfPastDays).toEqual(numberOfPastDays)
      })
    })
  })

  describe('data', () => {
    it('should set numberOfDays', () => {
      // When
      const wrapper = shallow(Calendar, { localVue, store })

      // Then
      expect(wrapper.vm.$data.numberOfDays).toEqual(8)
    })
    it('should set numberOfPastDays', () => {
      // When
      const wrapper = shallow(Calendar, { localVue, store })

      // Then
      expect(wrapper.vm.$data.numberOfPastDays).toEqual(2)
    })
  })

  describe('methods', () => {
    describe('formatDate', () => {
      it('should return formatted date with month and year', () => {
        // Given
        const date = moment('1992-11-07')
        const wrapper = shallow(Calendar, { localVue, store })

        // When
        const formattedDate = wrapper.vm.formatDate(date)

        // Then
        expect(formattedDate).toEqual('novembre 1992')
      })
    })

    xdescribe('showPreviousMonth', () => {
      describe('when current day is contained in the previous month', () => {
        it('should set current date to the same day of the previous month', () => {
          // Given
          const currentDate = moment('2017-10-30')
          const expectedPreviousMonth = '2017-09-30'
          const wrapper = shallow(Calendar, { localVue, store })
          wrapper.setData({ currentDate })

          // When
          wrapper.vm.showPreviousMonth()

          // Then
          expect(wrapper.vm.currentDate.format('YYYY-MM-DD')).toEqual(expectedPreviousMonth)
        })
      })

      describe('when current day is not contained in the previous month because it has less days', () => {
        it('should set current date to the last day of the previous month', () => {
          // Given
          const currentDate = moment('2017-10-31')
          const expectedPreviousMonthWithLessDays = '2017-09-30'
          const wrapper = shallow(Calendar, { localVue, store })
          wrapper.setData({ currentDate })

          // When
          wrapper.vm.showPreviousMonth()

          // Then
          expect(wrapper.vm.currentDate.format('YYYY-MM-DD')).toEqual(expectedPreviousMonthWithLessDays)
        })
      })
    })

    xdescribe('showNextMonth', () => {
      describe('when current day is contained in the next month', () => {
        it('should set current date to the same day of the next month', () => {
          // Given
          const currentDate = moment('2017-10-30')
          const expectedNextMonth = '2017-11-30'
          const wrapper = shallow(Calendar, { localVue, store })
          wrapper.setData({ currentDate })

          // When
          wrapper.vm.showNextMonth()

          // Then
          expect(wrapper.vm.currentDate.format('YYYY-MM-DD')).toEqual(expectedNextMonth)
        })
      })

      describe('when current day is not contained in the next month because it has less days', () => {
        it('should set current date to the last day of the next month', () => {
          // Given
          const currentDate = moment('2017-10-31')
          const expectedNextMonthWithLessDays = '2017-11-30'
          const wrapper = shallow(Calendar, { localVue, store })
          wrapper.setData({ currentDate })

          // When
          wrapper.vm.showNextMonth()

          // Then
          expect(wrapper.vm.currentDate.format('YYYY-MM-DD')).toEqual(expectedNextMonthWithLessDays)
        })
      })
    })
  })
})
