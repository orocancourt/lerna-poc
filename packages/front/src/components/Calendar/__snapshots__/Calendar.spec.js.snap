// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | Calendar template should match snapshot 1`] = `
<section
  id="calendar"
>
  <div
    class="calendar-navigation"
  >
    <button
      class="calendar-navigation-button btn btn--icon"
      type="button"
    >
      <div
        class="btn__content"
      >
        <i
          aria-hidden="true"
          class="icon material-icons"
        >
          keyboard_arrow_left
        </i>
      </div>
    </button>
     
    <h1>
      f√©vrier 2017
    </h1>
     
    <button
      class="calendar-navigation-button btn btn--icon"
      type="button"
    >
      <div
        class="btn__content"
      >
        <i
          aria-hidden="true"
          class="icon material-icons"
        >
          keyboard_arrow_right
        </i>
      </div>
    </button>
  </div>
   
  <div
    class="container table text-xs-center pa-0"
  >
    <!---->
     
    <!---->
  </div>
</section>
`;
